net/host.go

```diff
@@ -79,25 +107,45 @@ func (h *Host) P2pHost() P2pHost {
 	return h.h
 }
 
-// SetHandler sets the Handler instance used by the host.
-func (h *Host) SetHandler(handler Handler) {
-	fn := func() bool {
-		return len(handler.GetOffers()) == 0
+func (h *Host) advertisedNamespaces() []string {
+	provides := []string{""}
+
+	if len(h.makerHandler.GetOffers()) > 0 {
+		provides = append(provides, string(coins.ProvidesXMR))
 	}
 
-	h.handler = handler
-	h.h.SetShouldAdvertiseFunc(fn)
+	if h.isRelayer {
+		provides = append(provides, RelayerProvidesStr)
+	}
+
+	return provides
+}
+
+// SetHandlers sets the maker and taker instances used by the host.
+func (h *Host) SetHandlers(makerHandler MakerHandler, takerHandler TakerHandler) {
+	h.makerHandler = makerHandler
+	h.takerHandler = takerHandler
+	h.h.SetAdvertisedNamespacesFunc(h.advertisedNamespaces)
+
+	h.h.SetStreamHandler(queryProtocolID, h.handleQueryStream)
+	if h.isRelayer {
+		h.h.SetStreamHandler(relayProtocolID, h.handleRelayStream)
+	}
+	h.h.SetStreamHandler(swapID, h.handleProtocolStream)
```

Comment:
curious why the stream handlers were moved here instead of left in `Start()`? it doesn't make a huge difference, but this function was meant to be called specifically to set the maker/taker handlers. if you leave it here, update the comment

Reply:
As a general principle, methods named `Start` should avoid initialization that could have been done earlier during a constructor or other setup methods. This minimizes the time delta between invoking start and starting.  While there are no errors to handle in this particular case, pushing initialization code earlier, to when a program is likely to be in a single thread/process context, also leads to cleaner error handling. The method is called `SetHandlers`, and these are handlers, so it seemed like a good place to move the code. I'll update the comment.